### ISSUE TITLE
Training on TPU is broken by latest commits

### ISSUE THREAD

**Issue by InfiniteLife** [google/automl#523]
Running training now gives following error (model d3, device v3-8):

```
Parsing Inputs...
I0617 08:22:47.853078 139981562230592 efficientdet_arch.py:735] backbone+fpn+box params/flops = 11.904784M, 93.552447485B
I0617 08:22:47.853680 139981562230592 det_model_fn.py:99] LR schedule method: cosine
INFO:tensorflow:training_loop marked as finished
I0617 08:22:47.891298 139981562230592 error_handling.py:115] training_loop marked as finished
WARNING:tensorflow:Reraising captured error
W0617 08:22:47.891625 139981562230592 error_handling.py:149] Reraising captured error
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py", line 1654, in _create_c_op
    c_op = pywrap_tf_session.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimensions must be equal, but are 112 and 4 for '{{node focal_loss/mul_5}} = Mul[T=DT_FLOAT](focal_loss/mul_4, focal_loss/softmax_cross_entropy_with_logits/Reshape_2)' with input shapes: [4,112,112,18], [4,112,112].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "main.py", line 421, in <module>
    app.run(main)
  File "/usr/local/lib/python3.7/dist-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.7/dist-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "main.py", line 278, in main
    FLAGS.train_batch_size))
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3083, in train
    rendezvous.raise_errors()
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/tpu/error_handling.py", line 150, in raise_errors
    six.reraise(typ, value, traceback)
  File "/usr/local/lib/python3.7/dist-packages/six.py", line 703, in reraise
    raise value
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3078, in train
    saving_listeners=saving_listeners)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/estimator.py", line 349, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/estimator.py", line 1182, in _train_model
    return self._train_model_default(input_fn, hooks, saving_listeners)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/estimator.py", line 1211, in _train_model_default
    self.config)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 2915, in _call_model_fn
    config)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/estimator.py", line 1170, in _call_model_fn
    model_fn_results = self._model_fn(features=features, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3206, in _model_fn
    _train_on_tpu_system(ctx, model_fn_wrapper, dequeue_fn))
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3648, in _train_on_tpu_system
    device_assignment=ctx.device_assignment)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/tpu/tpu.py", line 1565, in split_compile_and_shard
    name=name)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/tpu/tpu.py", line 1280, in split_compile_and_replicate
    outputs = computation(*computation_inputs)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3634, in multi_tpu_train_steps_on_single_shard
    inputs=[0, _INITIAL_LOSS])
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/tpu/training_loop.py", line 178, in while_loop
    condition_wrapper, body_wrapper, inputs, name="", parallel_iterations=1)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/control_flow_ops.py", line 2688, in while_loop
    back_prop=back_prop)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/while_v2.py", line 196, in while_loop
    add_control_dependencies=add_control_dependencies)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py", line 981, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/while_v2.py", line 174, in wrapped_body
    outputs = body(*_pack_sequence_as(orig_loop_vars, args))
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/tpu/training_loop.py", line 121, in body_wrapper
    outputs = body(*(inputs + dequeue_ops))
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3633, in <lambda>
    lambda i, loss: [i + 1, single_tpu_train_step(i)],
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 1753, in train_step
    self._call_model_fn(features, labels))
  File "/usr/local/lib/python3.7/dist-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 2031, in _call_model_fn
    estimator_spec = self._model_fn(features=features, **kwargs)
  File "/mars/automl/efficientdet/det_model_fn.py", line 692, in efficientdet_model_fn
    variable_filter_fn=variable_filter_fn)
  File "/mars/automl/efficientdet/det_model_fn.py", line 470, in _model_fn
    cls_outputs, box_outputs, labels, params)
  File "/mars/automl/efficientdet/det_model_fn.py", line 263, in detection_loss
    normalizer=num_positives_sum)
  File "/mars/automl/efficientdet/det_model_fn.py", line 179, in focal_loss
    return alpha_factor * modulating_factor * ce / normalizer
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/math_ops.py", line 984, in binary_op_wrapper
    return func(x, y, name=name)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/math_ops.py", line 1283, in _mul_dispatch
    return gen_math_ops.mul(x, y, name=name)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/gen_math_ops.py", line 6092, in mul
    "Mul", x=x, y=y, name=name)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 744, in _apply_op_helper
    attrs=attr_protos, op_def=op_def)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py", line 595, in _create_op_internal
    compute_device)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py", line 3327, in _create_op_internal
    op_def=op_def)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py", line 1817, in __init__
    control_input_ops, op_def)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/ops.py", line 1657, in _create_c_op
    raise ValueError(str(e))
ValueError: Dimensions must be equal, but are 112 and 4 for '{{node focal_loss/mul_5}} = Mul[T=DT_FLOAT](focal_loss/mul_4, focal_loss/softmax_cross_entropy_with_logits/Reshape_2)' with input shapes: [4,112,112,18], [4,112,112].
```
I hard reset my revision to this commit https://github.com/google/automl/commit/c19585a2420a0d2f257b3c82292f411f3a6c699c#diff-50b32a4c6fac7eb8f22d7ec879a1a7d5 and on this commit training works fine.

**Comment by Steve-2040** [comment]
Hi,
I experience a similar error after using the latest efficientdet code.
Training using own data set on Windows with CPU only:

set CUDA_VISIBLE_DEVICES=""
python efficientdet/main.py --mode=train ^
    --training_file_pattern="model/stem_train.record" ^
    --model_name=efficientdet-d0 ^
    --model_dir=tmp/efficientdet-d0-scratch ^
    --train_batch_size=1 ^
    --num_examples_per_epoch=1587 --num_epochs=500 ^
    --hparams="num_classes=4,moving_average_decay=0,image_size=1280x1024,label_id_mapping=model/classes.yaml,"

W0617 23:21:02.267505 14344 error_handling.py:149] Reraising captured error
Traceback (most recent call last):
  File "C:\Users\Steve\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 1654, in _create_c_op
    c_op = pywrap_tf_session.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimensions must be equal, but are 160 and 128 for '{{node focal_loss/mul_5}} = Mul[T=DT_FLOAT](focal_loss/mul_4, focal_loss/softmax_cross_entropy_with_logits/Reshape_2)' with input shapes: [1,128,160,36], [1,128,160].

File "C:\Users\Steve\AppData\Local\Programs\Python\Python36\lib\site-packages\tensorflow\python\framework\ops.py", line 1657, in _create_c_op
    raise ValueError(str(e))
ValueError: Dimensions must be equal, but are 160 and 128 for '{{node focal_loss/mul_5}} = Mul[T=DT_FLOAT](focal_loss/mul_4, focal_loss/softmax_cross_entropy_with_logits/Reshape_2)' with input shapes: [1,128,160,36], [1,128,160].

Worked well before

**Comment by YumainOB** [comment]
The same problem is present in GPU training. The issue seems to be located in 
File "/home/obrousse/automl/efficientdet/det_model_fn.py", line 179, in focal_loss
    return alpha_factor * modulating_factor * ce / normalizer
in master.
It seeme that the sync4 brach is not affected by this issue and more problably that it does not include the bug.

**Comment by mingxingtan** [comment]
Fixed in #522
